metricflow-semantics/metricflow_semantics/model/semantics/linkable_element_set_base.py

```diff
@@ -0,0 +1,257 @@
+from metricflow_semantics.model.semantics.element_filter import LinkableElementFilter
+from metricflow_semantics.model.semantics.linkable_element import LinkableElementType
+from metricflow_semantics.specs.dimension_spec import DimensionSpec
+from metricflow_semantics.specs.entity_spec import EntitySpec
+from metricflow_semantics.specs.group_by_metric_spec import GroupByMetricSpec
+from metricflow_semantics.specs.instance_spec import LinkableInstanceSpec
+from metricflow_semantics.specs.patterns.spec_pattern import SpecPattern
+from metricflow_semantics.specs.time_dimension_spec import TimeDimensionSpec
+from metricflow_semantics.time.granularity import ExpandedTimeGranularity
+
+logger = logging.getLogger(__name__)
+
+
+class BaseLinkableElementSet(SemanticModelDerivation, ABC):
+    """Temporary interface used to migrate `LinkableElementSet`.
+
+    This class will be updated / renamed.
+    """
+
+    @abstractmethod
+    def filter(
+        self,
+        element_filter: LinkableElementFilter,
+    ) -> Self:
+        """Filter elements in the set.
+
+        First, only elements with at least one property in the "with_any_of" set are retained. Then, any elements with
+        a property in "without_any_of" set are removed. Lastly, any elements with all properties in without_all_of
+        are removed.
+        """
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def is_empty(self) -> bool:
+        """Whether this maps to an empty set of specs.
+
+        If `.specs` is created lazily, there are cases where the caller only needs to know that there are matching
+        specs.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def specs(self) -> Sequence[LinkableInstanceSpec]:
+        """Converts the items represented by this to the corresponding spec objects."""
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def annotated_specs(self) -> Sequence[AnnotatedSpec]:
+        """Return the annotated specs represented by this."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def filter_by_spec_patterns(self, spec_patterns: Sequence[SpecPattern]) -> Self:
+        """Filter the elements in the set by the given spec patters.
+
+        Returns a new set consisting of the elements in the `LinkableElementSet` that have a corresponding spec that
+        match all the given spec patterns.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def intersection(self, *others: Self) -> Self:
+        """Intersect this set with the other set to determine possible group-by items."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def union(self, *others: Self) -> Self:
+        """Intersect this set with the other set to determine possible group-by items."""
+        raise NotImplementedError
+
+
+@fast_frozen_dataclass()
+class AnnotatedSpec(SerializableDataclass):
```

Comment:
Is this basically just a generic `LinkableElement` that has all the possible attributes you might find on a `LinkableElement` for type convenience?

Reply:
That's about right.